File: weatherSDK/weather-sdk-java/src/main/java/com/example/weather/WeatherSDK.java
package com.example.weather;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import okhttp3.*;

import java.io.IOException;
import java.time.Duration;
import java.time.Instant;
import java.util.*;
import java.util.concurrent.*;

public class WeatherSDK {

  private static final String OPENWEATHER_URL = "https://api.openweathermap.org/data/2.5/weather";

  // Tracks active API keys to ensure no duplicate instance with the same key
  private static final Set<String> ACTIVE_KEYS = new HashSet<>();

  private final String apiKey;
  private final WeatherMode mode;
  private final ObjectMapper objectMapper;
  private final OkHttpClient httpClient;

  // Caching: city -> CacheEntry
  protected final Map<String, CacheEntry> cache = new ConcurrentHashMap<>();
  // Store city fetch order if we need to limit to 10
  final Deque<String> cityOrder = new LinkedList<>();

  // Executor for polling tasks
  private ScheduledExecutorService pollingExecutor;

  /**
   * Main constructor for the SDK.
   * 
   * @param apiKey The OpenWeather API key.
   * @param mode   The mode of SDK operation (ON_DEMAND or POLLING).
   * @throws WeatherSDKException if an SDK with the same API key already exists.
   */
  public WeatherSDK(String apiKey, WeatherMode mode) throws WeatherSDKException {
    // Ensure only one instance per API key
    synchronized (ACTIVE_KEYS) {
      if (ACTIVE_KEYS.contains(apiKey)) {
        throw new WeatherSDKException("An SDK with this API key already exists!");
      }
      ACTIVE_KEYS.add(apiKey);
    }

    this.apiKey = apiKey;
    this.mode = mode;
    this.objectMapper = new ObjectMapper();
    this.httpClient = new OkHttpClient();

    // If in polling mode, schedule periodic updates
    if (mode == WeatherMode.POLLING) {
      startPolling();
    }
  }

  /**
   * Retrieves the weather data for a given city in JSON format.
   * 
   * @param city The city name.
   * @return A {@link JsonNode} containing the weather data.
   * @throws WeatherSDKException if the city is invalid or network/API errors
   *                             occur.
   */
  public JsonNode getWeather(String city) throws WeatherSDKException {
    // Validate city name
    if (city == null || city.isEmpty()) {
      throw new WeatherSDKException("City name must not be empty");
    }

    // Check cache for data that is <10 minutes old
    CacheEntry entry = cache.get(city);
    if (entry != null &&
        Duration.between(entry.getTimestamp(), Instant.now()).toMinutes() < 10) {
      return entry.getData();
    }

    // Fetch fresh data
    return fetchAndCacheWeather(city);
  }

  /**
   * Internal method to fetch data from the OpenWeather API and store in cache.
   */
  private JsonNode fetchAndCacheWeather(String city) throws WeatherSDKException {
    Request request = new Request.Builder()
        .url(OPENWEATHER_URL + "?q=" + city + "&appid=" + apiKey + "&units=metric")
        .build();

    try (Response response = httpClient.newCall(request).execute()) {
      if (!response.isSuccessful()) {
        throw new WeatherSDKException("Failed to fetch weather: " + response.code());
      }
      String body = response.body().string();
      JsonNode node = objectMapper.readTree(body);

      // Cache it
      cacheWeather(city, node);
      return node;
    } catch (IOException e) {
      throw new WeatherSDKException("Network error: " + e.getMessage());
    }
  }

  /**
   * Caches the provided weather data for the city, respecting the 10-city limit.
   */
  private void cacheWeather(String city, JsonNode data) {
    // If city is new and we already have 10 stored, remove the oldest
    if (!cache.containsKey(city) && cache.size() >= 10) {
      String oldest = cityOrder.pollFirst(); // remove first
      if (oldest != null) {
        cache.remove(oldest);
      }
    }
    // If city already exists, remove from queue so we can place it at the end
    cityOrder.remove(city);

    // Insert at the end
    cityOrder.addLast(city);
    cache.put(city, new CacheEntry(data, Instant.now()));
  }

  /**
   * Starts a polling executor that periodically refreshes weather data for all
   * cached cities.
   */
  private void startPolling() {
    pollingExecutor = Executors.newSingleThreadScheduledExecutor();
    // Poll every 5 minutes
    pollingExecutor.scheduleAtFixedRate(() -> {
      for (String city : cache.keySet()) {
        try {
          fetchAndCacheWeather(city);
        } catch (WeatherSDKException e) {
          // Log or handle if needed
        }
      }
    }, 0, 5, TimeUnit.MINUTES);
  }

  /**
   * Deletes the SDK instance: stops polling, clears caches, and frees up the API
   * key.
   */
  public void delete() {
    if (pollingExecutor != null && !pollingExecutor.isShutdown()) {
      pollingExecutor.shutdownNow();
    }
    synchronized (ACTIVE_KEYS) {
      ACTIVE_KEYS.remove(apiKey);
    }
    cache.clear();
    cityOrder.clear();
  }
}

File: weatherSDK/weather-sdk-python/weather_sdk/weather_sdk.py
import threading
import time
from collections import OrderedDict

import requests

from .exceptions import WeatherSDKException
from .weather_mode import WeatherMode

_ACTIVE_KEYS = set()


class WeatherSDK:
    OPENWEATHER_URL = "https://api.openweathermap.org/data/2.5/weather"

    def __init__(self, api_key: str, mode: WeatherMode):
        if api_key in _ACTIVE_KEYS:
            raise WeatherSDKException("An SDK with this API key already exists!")
        _ACTIVE_KEYS.add(api_key)

        self._api_key = api_key
        self._mode = mode
        self._cache = OrderedDict()  # city -> (data, timestamp)
        self._lock = threading.Lock()

        if self._mode == WeatherMode.POLLING:
            self._polling_thread = threading.Thread(
                target=self._polling_loop, daemon=True
            )
            self._polling_thread.start()
        else:
            self._polling_thread = None

    def get_weather(self, city: str) -> dict:
        if not city:
            raise WeatherSDKException("City name must not be empty")

        with self._lock:
            if city in self._cache:
                data, ts = self._cache[city]
                if (time.time() - ts) < 600:  # Cache valid for 10 min
                    return data

        return self._fetch_and_cache_weather(city)

    def _fetch_and_cache_weather(self, city: str) -> dict:
        params = {"q": city, "appid": self._api_key, "units": "metric"}
        try:
            resp = requests.get(self.OPENWEATHER_URL, params=params)
            if resp.status_code == 401:
                raise WeatherSDKException("Invalid API Key. Please check your API key.")
            elif resp.status_code == 404:
                raise WeatherSDKException(f"City '{city}' not found.")
            elif not resp.ok:
                raise WeatherSDKException(
                    f"Failed to fetch weather: HTTP {resp.status_code} {resp.text}"
                )

            data = resp.json()
            with self._lock:
                self._cache_weather(city, data)
            return data

        except requests.exceptions.RequestException as e:
            raise WeatherSDKException(f"Network error: {e}")

    def _cache_weather(self, city: str, data: dict):
        if city not in self._cache and len(self._cache) >= 10:
            self._cache.popitem(last=False)

        if city in self._cache:
            del self._cache[city]

        self._cache[city] = (data, time.time())

    def _polling_loop(self):
        while True:
            with self._lock:
                for city in list(self._cache.keys()):
                    try:
                        self._fetch_and_cache_weather(city)
                    except WeatherSDKException:
                        pass
            time.sleep(300)

    def delete(self):
        if self._api_key in _ACTIVE_KEYS:
            _ACTIVE_KEYS.remove(self._api_key)
        with self._lock:
            self._cache.clear()
